File: csharp/843.guess-the-word.cs
/*
 * @lc app=leetcode id=843 lang=csharp
 *
 * [843] Guess the Word
 */

using System.Linq;

// @lc code=start
/**
 * // This is the Master's API interface.
 * // You should not implement it, or speculate about its implementation
 * class Master {
 *     public int Guess(string word);
 * }
 */
class Solution {
    public void FindSecretWord(string[] wordlist, Master master) {
        int count = 0;
        while (count != 6) {
            var guess = wordlist[wordlist.Length / 2];
            count = master.Guess(guess);
            wordlist = wordlist.Where(word => word.Zip(guess, (c1, c2) => c1 == c2).Where(val => val).Count() == count).ToArray();
        }
    }
}
// @lc code=end

File: java/843.guess-the-word.java
/*
 * @lc app=leetcode id=843 lang=java
 *
 * [843] Guess the Word
 */

// @lc code=start
/**
 * // This is the Master's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface Master {
 *     public int guess(String word) {}
 * }
 */
class Solution {
    public void findSecretWord(String[] wordlist, Master master) {
        while (true) {
            String guess = wordlist[wordlist.length / 2];
            int count = master.guess(guess);
            wordlist = Arrays.stream(wordlist).filter(w -> {
                int t = 0;
                for (int i = 0; i < 6; ++i) {
                    if (guess.charAt(i) == w.charAt(i)) {
                        ++t;
                    }
                }
                return t == count;
            }).toArray(String[]::new);
            if (count == 6) {
                break;
            }
        }
    }
}
// @lc code=end

